locale.c

```diff
@@ -768,6 +768,200 @@ Perl_locale_panic(const char * msg,
 #define setlocale_failure_panic_c(cat, cur, fail, line, higher_line)        \
    setlocale_failure_panic_i(cat##_INDEX_, cur, fail, line, higher_line)
 
+#if defined(USE_POSIX_2008_LOCALE) && defined(LC_ALL)
+
+STATIC parse_LC_ALL_string_return
+S_parse_LC_ALL_string(pTHX_ const char * string,
+                            const char ** output,
+                            const line_t caller_line)
+{
+    /* This function parses the value of the input 'string' which is expected
+     * to be the representation of an LC_ALL locale, and splits the result into
+     * the values for the individual component categories, returning those in
+     * the 'output' array.  Each array value will be a savepv() copy that is
+     * the responsibility of the caller to make sure gets freed
+     *
+     * The locale for each category is independent of the other categories.
+     * Often, they are all the same, but certainly not always.  Perl, in fact,
+     * usually keeps LC_NUMERIC in the C locale, regardless of the underlying
+     * locale.  LC_ALL has to be able to represent the case of when not all
+     * categories have the same locale.  Platforms have differing ways of
+     * representing this.  Internally, this file uses the 'name=value;'
+     * representation found on some platforms, so this function always looks
+     * for and parses that.
+     *
+     * Often, all categories will have the same locale.  In that case, the
+     * input 'string' likely is a single value, and no splitting is needed.
+     * In such cases, this function doesn't store anything into 'output', and
+     * returns 'no_array'.
+     *
+     * Otherwise, the input 'string' may not be valid.  This function looks
+     * mainly for syntactic errors, and if found, returns 'invalid'.  'output'
+     * will not be filled in in that case, but the input state of it isn't
+     * necessarily preserved.  Turning on -DL debugging will give details as to
+     * the error.
+     *
+     * Otherwise, output[] will be filled with the individual locale names for
+     * all categories on the system, 'full_array' will be returned, and the
+     * caller needs to arrange for each to be freed.
+     *
+     */
+
+    DEBUG_Lv(PerlIO_printf(Perl_debug_log,
+                           "Entering parse_LC_ALL_string; called from %"    \
+                           LINE_Tf "\nnew='%s'\n", caller_line, string));
+
+    const char separator[] = ";";
+    const Size_t separator_len = 1;
```

Comment:
Are these ever not ";"? Or not a single character?

I do see this has changed from the code it replaces which uses character checks for ';'.

Reply:
A future commit will take advantage of this being a variable, and with a different value.

I have not seen any instances of more than a single character, but it seemed less brittle to allow a string for any such platforms, and with no extra logic needed.